### ISSUE TITLE
Question: How to correctly implement identity delegation in microservices?

### ISSUE THREAD

**Issue by aurelienshz** [panva/node-oidc-provider#211]
Hi! :wave: 

**tl;dr** : Is exchanging an access token against a JWT at the edge of the API layer a correct way of delegating identity through microservices?

We are using node-oidc-provider to manage identities through several microservices, with an API gateway in front of them to distribute requests to the right service and aggregate responses. We will be consuming this API layer through several front-end SPAs, and we have implemented implicit grant flows on these SPAs to be able to access the user's identity (using the id token returned by the OIDC provider).

We are passing the access_token to the requests fired to the API Gateway in order to authenticate them (and we keep the id_token internal to the SPAs and use it only to customize the UX to the currently logged in user).

We are wondering about the best way to distribute identity among microservices so that they can have access to the user's identity & authorities, and read token scopes, in order to make decisions regarding authorizations.

Here are the four possible options we came up with : 

1. have the API gateway use the token introspection or the userinfo endpoint to retrieve the user's info, and then generate and sign its own JWTs representing the user's identity (the underlying microservices would then have access to the user's identity through the passed JWT and use the API gateway's public key to verify them) : sounds pretty complex, and redundant, since we already have a service somewhere dedicated to signing.

2. have the API gateway make all the authorization decisions, then hit all the underlying microservices with an all-access service account (is this the right terminology?) : might be easier to implement, but less auditable, less secure (if the service account is compromised, everything's open), and more convoluted because the API gateway might have to fire several requests just to make the authorization decisions (not to mention we would still have to somehow find a way to pass the user's identity to the microservices in some cases where at least a reference to the end user's account is needed) ;

3. move from implicit grant to the authorization code grant implemented on the API gateway (the client would then be the API gateway instead of the front-end SPAs) : in that case, the API layer would now have immediate access to an id_token, but the problem would be reversed: how would the identity of the user be passed back to the front-end applications ?

4. have the API gateway contact the OIDC provider to exchange the access token given by the client SPA against a JWT (pretty much like an internal id_token), and add this JWT to the requests to the underlying microservices : this sounds like the most elegant option, but maybe I'm completely wrong ?

The last option seems to be the one proposed [here](https://nordicapis.com/how-to-control-user-identity-within-microservices/#letallmicroservicesconsumejwt) or [here](https://medium.com/@robert.broeckelmann/identity-propagation-in-an-api-gateway-architecture-c0f9bbe9273b), and there is even an [IETF draft](https://tools.ietf.org/html/draft-ietf-oauth-token-exchange-12) about a token exchange flow on the token endpoint.

I'd like to add that, apart from the API gateway and the OIDC provider, our microservices are not exposed on the scary internet. That's why we feel like it would be secure enough to pass an id_token around between them (whereas passing an id_token obtained by an SPA in each and every HTTP request, is, in my understanding, considered bad practice).

We would be happy to hear any opinions regarding this question, and, if we finally go with the token exchange approach, we would love to hear from the maintainers of this project about the right way to implement it. We would also love to provide our implementation as a PR or as an example app in case this question arises again!

**Comment by panva** [comment]
Hi @aurelienshz, I believe you're better off asking these questions in the IETF or OpenID WG mailing lists.

> if we finally go with the token exchange approach, we would love to hear from the maintainers of this project about the right way to implement it

Registering a custom grant type > see doc https://github.com/panva/node-oidc-provider/blob/master/docs/configuration.md#custom-grant-types

**Comment by aurelienshz** [comment]
Hi! Thanks a lot for your very quick reply, your suggestion of heading to the IETF mailing lists, and thanks for closing the issue. We're going to try and implement the draft I linked in my question.If that works well, would you be interested in a PR for this as an experimental, opt-in feature, or do you feel like that would fall out of scope of this project?

**Comment by panva** [comment]
I feel like it’s going to end up littered with custom policies specific to your IdP, nevertheless i’d be interested in the solution and if along the way you feel like the lib doesn’t have the tools to achieve this let me know.
